src/coreclr/jit/hwintrinsicxarch.cpp

```diff
@@ -2877,33 +2877,27 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
 
         case NI_Vector128_Sum:
         case NI_Vector256_Sum:
+        case NI_Vector512_Sum:
         {
             assert(sig->numArgs == 1);
             var_types simdType = getSIMDTypeForSize(simdSize);
 
             if ((simdSize == 32) && !compOpportunisticallyDependsOn(InstructionSet_AVX2))
             {
-                // Vector256 for integer types requires AVX2
+                // Vector256 requires AVX2
                 break;
             }
-            else if (varTypeIsFloating(simdBaseType))
+            else if ((simdSize == 16) && !compOpportunisticallyDependsOn(InstructionSet_SSE2))
             {
-                if (!compOpportunisticallyDependsOn(InstructionSet_SSE3))
-                {
-                    // Floating-point types require SSE3.HorizontalAdd
-                    break;
-                }
-            }
-            else if (!compOpportunisticallyDependsOn(InstructionSet_SSSE3))
-            {
-                // Integral types require SSSE3.HorizontalAdd
                 break;
             }
-            else if (varTypeIsByte(simdBaseType) || varTypeIsLong(simdBaseType))
+#if defined(TARGET_X86)
+            else if (varTypeIsLong(simdBaseType))
```

Comment:
What support is missing for `long` on 32-bit?

I would have expected this to generally "just work" and for us to be able to use `_ToScalar` provided `SSE4.1` is supported given that `GetElement` has the required decomposition support for that case.

Reply:
I had a failure with decomposition. I'll try again and get the exact details

Reply:
@tannergooding It's passing now

Reply:
![image](https://github.com/dotnet/runtime/assets/17969209/1b093794-3e42-4c56-aecd-f5078f29b924)

I need to debug this to pin point exactly what's going wrong here.

Reply:
Not sure if this is the right way to handle this - https://github.com/dotnet/runtime/pull/95568/commits/9e3657f216cc2491ff1c0cb057a8214ad3bdba71

Reply:
That won't quite work unfortunately and the actual handling is going to be a little bit more complex.

Noting that I'm fine with this being handled as a separate PR, I had initially thought it might be slightly simpler but looks like its not quite all there..

The "proper" fix likely entails:
1. Extracting this logic to a `gtNewSimdToScalarNode` helper: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsicxarch.cpp#L2917-L2937
2. Updating various places that are manually doing `gtNewSimdHWIntrinsicNode(retType, op1, NI_Vector###_ToScalar, simdBaseJitType, simdSize)` to call the introduced helper
3. Have the `impSpecialIntrinsic` handling for `NI_Vector###_Sum` do:
```cpp
#if defined(TARGET_X86)
            else if (varTypeIsLong(simdBaseType) && !compOpportunisticallyDependsOn(InstructionSet_SSE41))
            {
                // We need SSE41 to handle long, use software fallback
                break;
            }
#endif // TARGET_X86
```

At some future point we can ensure that `NI_Vector###_GetElement` has decomp handling for pre-SSE4.1 as well, it's just slightly more complex given that `TYP_INT` currently requires `SSE4.1` as well. The fix to get that working is likely to just reuse the `TYP_FLOAT` handling for getting the relevant 32-bit part and then using `ToScalar` (shifting can then be used for the relevant 8-bit or 16-bit part for byte/short).

Reply:
I've updated the implementation to use the helper. Waiting for CI tests to pass. My local tests passed